In [ ]:
import os

from google.colab import drive
drive.mount('/content/drive')
os.chdir("/content/drive/My Drive/DATA SCIENCE/mòdul 6")
os.listdir()

Mounted at /content/drive


['tec_mbdds_taa_1.pdf',
 'tec_mbdds_taa_2.pdf',
 '3. Naive Bayes',
 '1. Árboles decisión',
 '2. KNN',
 '4. Random forest',
 '5. Gradient boosting',
 'M6_AG1_Grupo11.ipynb',
 'act_grupal_News_Category_Dataset_v2.json',
 'Copy of M6_AG1_Grupo11.ipynb',
 '6. ada boost',
 'HM_interactions.csv.csv']

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
df = pd.read_csv('HM_interactions.csv.csv', sep=',', usecols=['customer_id', 'article_id'])

In [ ]:
df = pd.DataFrame(df)
df.head

<bound method NDFrame.head of         customer_id  article_id
0             10054        5498
1              8612        4092
2             12973        4109
3              9820       12918
4              3329        1739
...             ...         ...
382317        12873       18345
382318         1010       11221
382319         8493       13473
382320         7232        4383
382321         5460       13815

[382322 rows x 2 columns]>

In [ ]:
df['Count'] = 1
df.head

<bound method NDFrame.head of         customer_id  article_id  Count
0             10054        5498      1
1              8612        4092      1
2             12973        4109      1
3              9820       12918      1
4              3329        1739      1
...             ...         ...    ...
382317        12873       18345      1
382318         1010       11221      1
382319         8493       13473      1
382320         7232        4383      1
382321         5460       13815      1

[382322 rows x 3 columns]>

In [ ]:
df.nunique()

customer_id    13359
article_id     21475
Count              1
dtype: int64

In [ ]:
df.isnull().any()

customer_id    False
article_id     False
Count          False
dtype: bool

No hay ningún missing value

In [ ]:
df[df.duplicated(keep='last')]
df.shape

(382322, 3)

In [ ]:
df1 = df.drop_duplicates()
df1.shape

(351760, 3)

Seleccionamos top articulos

Sabemos que hay 219 artículos que se han comprado más de 100 veces. 

Ahora queremos que nuestro dataset df1 contenga solo aquellos artículos comprados más de 100 veces, esto es, solo los 219 artículos. 

In [ ]:
df_filter = df1.loc[seleccion_articulos]
df_filter

customer_id  article_id  Count
article_id                                
32                 3701        9220      1
132               10905       19495      1
221               11420       14764      1
240               10478       19721      1
241                1436         973      1
...                 ...         ...    ...
18586              3788        6671      1
19668               213       20009      1
20814              1304        2593      1
20845              3728       17340      1
20913             12669       14194      1

[219 rows x 3 columns]

In [ ]:
df1.shape

(351760, 3)

In [ ]:
#df_p = df1.loc[seleccion_articulos, ['customer_id', 'Count']]

df2 = df1.groupby('article_id').count()
df2.head()

customer_id  Count
article_id                    
0                     3      3
1                    10     10
2                     3      3
3                    25     25
4                    27     27

In [ ]:
df3 = df2[df2['Count']>100]
df3

customer_id  Count
article_id                    
32                  111    111
132                 108    108
221                 128    128
240                 188    188
241                 163    163
...                 ...    ...
18586               205    205
19668               103    103
20814               122    122
20845               126    126
20913               101    101

[219 rows x 2 columns]

In [ ]:
#df1 es la original

df4 = df1.merge(df3,on=['article_id'])
df4

customer_id_x  article_id  Count_x  customer_id_y  Count_y
0               8612        4092        1            147      147
1              13089        4092        1            147      147
2               2413        4092        1            147      147
3               1920        4092        1            147      147
4               5015        4092        1            147      147
...              ...         ...      ...            ...      ...
30646           9113        3161        1            103      103
30647           2190        3161        1            103      103
30648           7710        3161        1            103      103
30649           2581        3161        1            103      103
30650           4109        3161        1            103      103

[30651 rows x 5 columns]

La matrix solo nos incluirá los custmers que han comprado al menos uno de los productos que se han comprado al menos 100 veces

Renombramos las columnas y nos quedamos solo con customer_id_x y article_id	

In [ ]:
df5 = df4.drop(columns= ['customer_id_y','Count_y'])
df5 = df5.rename(columns={'customer_id_x':'customer_id', 'Count_x':'count'})
df5

customer_id  article_id  count
0             8612        4092      1
1            13089        4092      1
2             2413        4092      1
3             1920        4092      1
4             5015        4092      1
...            ...         ...    ...
30646         9113        3161      1
30647         2190        3161      1
30648         7710        3161      1
30649         2581        3161      1
30650         4109        3161      1

[30651 rows x 3 columns]

Ahora creamos el matrix

In [ ]:
matrix = df5.pivot_table(
  index='customer_id',
  columns='article_id',
  values='count',
  aggfunc='sum'
)

matrix

article_id   32     132    221    240    241    420    421    422    495    \
customer_id                                                                  
0              NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1              NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3              NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4              NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
5              NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
...            ...    ...    ...    ...    ...    ...    ...    ...    ...   
13353          NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13354          NaN    NaN    1.0    NaN    NaN    NaN    NaN    NaN    1.0   
13355          NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13356          NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13358          NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

article_id   847    ...  18321  18349  18549  18579  18585  18586  19668  \
customer_id         ...                                                    
0              NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1              NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3              NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4              NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
5              NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
...            ...  ...    ...    ...    ...    ...    ...    ...    ...   
13353          NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13354          NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13355          NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13356          NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13358          NaN  ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

article_id   20814  20845  20913  
customer_id                       
0              NaN    NaN    NaN  
1              NaN    NaN    NaN  
3              NaN    NaN    NaN  
4              NaN    NaN    NaN  
5              NaN    NaN    NaN  
...            ...    ...    ...  
13353          NaN    NaN    NaN  
13354          NaN    NaN    NaN  
13355          NaN    NaN    NaN  
13356          NaN    NaN    NaN  
13358          NaN    NaN    NaN  

[11079 rows x 219 columns]

A continuación lo pasamos a binario

In [ ]:
matrix2 = matrix.applymap(lambda x:1 if x>0 else 0)
matrix2

article_id   32     132    221    240    241    420    421    422    495    \
customer_id                                                                  
0                0      0      0      0      0      0      0      0      0   
1                0      0      0      0      0      0      0      0      0   
3                0      0      0      0      0      0      0      0      0   
4                0      0      0      0      0      0      0      0      0   
5                0      0      0      0      0      0      0      0      0   
...            ...    ...    ...    ...    ...    ...    ...    ...    ...   
13353            0      0      0      0      0      0      0      0      0   
13354            0      0      1      0      0      0      0      0      1   
13355            0      0      0      0      0      0      0      0      0   
13356            0      0      0      0      0      0      0      0      0   
13358            0      0      0      0      0      0      0      0      0   

article_id   847    ...  18321  18349  18549  18579  18585  18586  19668  \
customer_id         ...                                                    
0                0  ...      0      0      0      0      0      0      0   
1                0  ...      0      0      0      0      0      0      0   
3                0  ...      0      0      0      0      0      0      0   
4                0  ...      0      0      0      0      0      0      0   
5                0  ...      0      0      0      0      0      0      0   
...            ...  ...    ...    ...    ...    ...    ...    ...    ...   
13353            0  ...      0      0      0      0      0      0      0   
13354            0  ...      0      0      0      0      0      0      0   
13355            0  ...      0      0      0      0      0      0      0   
13356            0  ...      0      0      0      0      0      0      0   
13358            0  ...      0      0      0      0      0      0      0   

article_id   20814  20845  20913  
customer_id                       
0                0      0      0  
1                0      0      0  
3                0      0      0  
4                0      0      0  
5                0      0      0  
...            ...    ...    ...  
13353            0      0      0  
13354            0      0      0  
13355            0      0      0  
13356            0      0      0  
13358            0      0      0  

[11079 rows x 219 columns]

Creación modelo colaborativo

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
matrix3_user_user = pd.DataFrame(cosine_similarity(matrix2))
matrix3_user_user 

0      1      2      3      4      5      6      7         8      \
0        1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
1        0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
2        0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0  0.000000   
3        0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0  0.000000   
4        0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0  0.000000   
...      ...    ...    ...    ...    ...    ...    ...    ...       ...   
11074    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
11075    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
11076    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.288675   
11077    0.5    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
11078    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   

          9      ...  11069     11070     11071  11072     11073  11074  \
0      0.000000  ...    0.0  0.000000  0.000000    0.0  0.408248    0.0   
1      0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
2      0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
3      0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
4      0.000000  ...    0.0  0.288675  0.000000    0.0  0.000000    0.0   
...         ...  ...    ...       ...       ...    ...       ...    ...   
11074  0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    1.0   
11075  0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
11076  0.000000  ...    0.0  0.288675  0.000000    0.0  0.000000    0.0   
11077  0.000000  ...    0.0  0.000000  0.353553    0.0  0.408248    0.0   
11078  0.258199  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   

       11075     11076  11077     11078  
0        0.0  0.000000    0.5  0.000000  
1        0.0  0.000000    0.0  0.000000  
2        0.0  0.000000    0.0  0.000000  
3        0.0  0.000000    0.0  0.000000  
4        0.0  0.000000    0.0  0.000000  
...      ...       ...    ...       ...  
11074    0.0  0.000000    0.0  0.000000  
11075    1.0  0.000000    0.0  0.000000  
11076    0.0  1.000000    0.0  0.288675  
11077    0.0  0.000000    1.0  0.000000  
11078    0.0  0.288675    0.0  1.000000  

[11079 rows x 11079 columns]

Ahora creamos un df con el resultado de esta matrix y lo guardamos poniéndole un índice

In [ ]:
matrix3_user_user.columns= matrix2.index
matrix3_user_user 

customer_id  0      1      3      4      5      6      7      9         10     \
0              1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
1              0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
2              0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0  0.000000   
3              0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0  0.000000   
4              0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0  0.000000   
...            ...    ...    ...    ...    ...    ...    ...    ...       ...   
11074          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
11075          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
11076          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.288675   
11077          0.5    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
11078          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   

customer_id     11     ...  13346     13347     13348  13349     13351  13353  \
0            0.000000  ...    0.0  0.000000  0.000000    0.0  0.408248    0.0   
1            0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
2            0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
3            0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
4            0.000000  ...    0.0  0.288675  0.000000    0.0  0.000000    0.0   
...               ...  ...    ...       ...       ...    ...       ...    ...   
11074        0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    1.0   
11075        0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
11076        0.000000  ...    0.0  0.288675  0.000000    0.0  0.000000    0.0   
11077        0.000000  ...    0.0  0.000000  0.353553    0.0  0.408248    0.0   
11078        0.258199  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   

customer_id  13354     13355  13356     13358  
0              0.0  0.000000    0.5  0.000000  
1              0.0  0.000000    0.0  0.000000  
2              0.0  0.000000    0.0  0.000000  
3              0.0  0.000000    0.0  0.000000  
4              0.0  0.000000    0.0  0.000000  
...            ...       ...    ...       ...  
11074          0.0  0.000000    0.0  0.000000  
11075          1.0  0.000000    0.0  0.000000  
11076          0.0  1.000000    0.0  0.288675  
11077          0.0  0.000000    1.0  0.000000  
11078          0.0  0.288675    0.0  1.000000  

[11079 rows x 11079 columns]

In [ ]:
matrix3_user_user['customer_id'] = matrix2.index
matrix3_user_user = matrix3_user_user.set_index('customer_id')
matrix3_user_user

customer_id  0      1      3      4      5      6      7      9         10     \
customer_id                                                                     
0              1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
1              0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
3              0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0  0.000000   
4              0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0  0.000000   
5              0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0  0.000000   
...            ...    ...    ...    ...    ...    ...    ...    ...       ...   
13353          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
13354          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
13355          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.288675   
13356          0.5    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   
13358          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.000000   

customer_id     11     ...  13346     13347     13348  13349     13351  13353  \
customer_id            ...                                                      
0            0.000000  ...    0.0  0.000000  0.000000    0.0  0.408248    0.0   
1            0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
3            0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
4            0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
5            0.000000  ...    0.0  0.288675  0.000000    0.0  0.000000    0.0   
...               ...  ...    ...       ...       ...    ...       ...    ...   
13353        0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    1.0   
13354        0.000000  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   
13355        0.000000  ...    0.0  0.288675  0.000000    0.0  0.000000    0.0   
13356        0.000000  ...    0.0  0.000000  0.353553    0.0  0.408248    0.0   
13358        0.258199  ...    0.0  0.000000  0.000000    0.0  0.000000    0.0   

customer_id  13354     13355  13356     13358  
customer_id                                    
0              0.0  0.000000    0.5  0.000000  
1              0.0  0.000000    0.0  0.000000  
3              0.0  0.000000    0.0  0.000000  
4              0.0  0.000000    0.0  0.000000  
5              0.0  0.000000    0.0  0.000000  
...            ...       ...    ...       ...  
13353          0.0  0.000000    0.0  0.000000  
13354          1.0  0.000000    0.0  0.000000  
13355          0.0  1.000000    0.0  0.288675  
13356          0.0  0.000000    1.0  0.000000  
13358          0.0  0.288675    0.0  1.000000  

[11079 rows x 11079 columns]

Cogemos los customers con más similitud

In [ ]:
matrix3_user_user.loc[12350.0].sort_values(ascending=False)

customer_id
12350    1.000000
3085     0.474342
11549    0.447214
12855    0.447214
2900     0.447214
           ...   
4693     0.000000
4694     0.000000
4698     0.000000
4699     0.000000
13358    0.000000
Name: 12350, Length: 11079, dtype: float64

A continuación, vemos los top clientes con más similitud con el cliente 12350. Por ejemplo el cliente 3085 es el que tiene más similitud con un valor de 0.47

Vamos a ver qué ha comprado el cliente 12350

In [ ]:
items_comprador_A = set(matrix2.loc[12350.0].iloc[
    matrix2.loc[12350.0].to_numpy().nonzero()
].index)

items_comprador_A

{32, 221, 1550, 1654, 3658, 4377, 6287, 9256, 11901, 12818}

Estos son los productos comprados por el comprador 12350 A. Ahora vamos a ver los que ha comprado el comprador con más similtud que A, que en este caso es 3085

In [ ]:
items_comprador_B = set(matrix2.loc[3085].iloc[
    matrix2.loc[3085].to_numpy().nonzero()
].index)

items_comprador_B

{1110, 1550, 4377, 11901}

Ahora miraremos qué items ha comprado A que el comprador B todavía no

In [ ]:
items_a_recomendar_a_B = items_comprador_A - items_comprador_B
items_a_recomendar_a_B

{32, 221, 1654, 3658, 6287, 9256, 12818}

Estos son los productos que podríamos recomendar a B